contrib/benchmarkDictBuilder/benchmark.c

```diff
@@ -0,0 +1,458 @@
+  }
+
+  /* Calculate offset for each sample */
+  offsets = (size_t *)malloc((srcInfo->nbSamples + 1) * sizeof(size_t));
+  offsets[0] = 0;
+  for (i = 1; i <= srcInfo->nbSamples; i++) {
+    offsets[i] = offsets[i - 1] + srcInfo->samplesSizes[i - 1];
+  }
+
+  /* Compress each sample and sum their sizes*/
+  const BYTE *const samples = (const BYTE *)srcInfo->srcBuffer;
+  for (i = 0; i < srcInfo->nbSamples; i++) {
+    const size_t compressedSize = ZSTD_compress_usingCDict(cctx, dst, dstCapacity, samples + offsets[i], srcInfo->samplesSizes[i], cdict);
+    if (ZSTD_isError(compressedSize)) {
+      cRatio = -1;
+      goto _cleanup;
+    }
+    totalCompressedSize += compressedSize;
+  }
+
+  /* Sum orignal sizes */
+  for (i = 0; i<srcInfo->nbSamples; i++) {
+    totalOriginalSize += srcInfo->samplesSizes[i];
+  }
+
+  /* Calculate compression ratio */
+  DISPLAYLEVEL(2, "original size is %lu\n", totalOriginalSize);
+  DISPLAYLEVEL(2, "compressed size is %lu\n", totalCompressedSize);
+  cRatio = (double)totalOriginalSize/(double)totalCompressedSize;
+
+_cleanup:
+  if(dst) {
+    free(dst);
+  }
+  if(offsets) {
+    free(offsets);
+  }
+  ZSTD_freeCCtx(cctx);
+  ZSTD_freeCDict(cdict);
+  return cRatio;
+}
+
+
+/** FreeDictInfo() :
+ *  Free memory allocated for dictInfo
+ */
+void freeDictInfo(dictInfo* info) {
+  if (!info) return;
+  if (info->dictBuffer) free((void*)(info->dictBuffer));
+  free(info);
+}
+
+
+
+/*-********************************************************
+  *  Benchmarking functions
+**********************************************************/
+/** benchmarkRandom() :
+ *  Measure how long random dictionary builder takes and compression ratio with the random dictionary
+ *  @return 0 if benchmark successfully, 1 otherwise
+ */
+int benchmarkRandom(sampleInfo *srcInfo, unsigned maxDictSize, ZDICT_random_params_t *randomParam) {
+  const int displayLevel = randomParam->zParams.notificationLevel;
+  int result = 0;
+  clock_t t;
+  t = clock();
+  dictInfo* dInfo = createDictFromFiles(srcInfo, maxDictSize, randomParam, NULL, NULL);
+  t = clock() - t;
+  double time_taken = ((double)t)/CLOCKS_PER_SEC;
+  if (!dInfo) {
+    DISPLAYLEVEL(1, "RANDOM does not train successfully\n");
+    result = 1;
+    goto _cleanup;
+  }
+  DISPLAYLEVEL(2, "RANDOM took %f seconds to execute \n", time_taken);
+
+  double cRatio = compressWithDict(srcInfo, dInfo, randomParam->zParams.compressionLevel, displayLevel);
+  if (cRatio < 0) {
+    DISPLAYLEVEL(1, "Compressing with RANDOM dictionary does not work\n");
+    result = 1;
+    goto _cleanup;
+  }
+  DISPLAYLEVEL(2, "Compression ratio with random dictionary is %f\n", cRatio);
+
+
+_cleanup:
+  freeDictInfo(dInfo);
+  return result;
+}
+
+/** benchmarkCover() :
+ *  Measure how long random dictionary builder takes and compression ratio with the cover dictionary
+ *  @return 0 if benchmark successfully, 1 otherwise
+ */
+int benchmarkCover(sampleInfo *srcInfo, unsigned maxDictSize,
```

Comment:
Lets merge these three functions, like you did with `createDictFromFiles()`.

You can derive the name to print from which parameter is non-null, or just pass it as an argument.